File: lib-segment/src/main/java/websiteschema/mpsegment/graph/BigramDijkstra.java
package websiteschema.mpsegment.graph;

import websiteschema.mpsegment.dict.IWord;

public class BigramDijkstra extends DijkstraImpl {

    private WordBigram wordBigram;

    public BigramDijkstra(WordBigram wordBigram) {
        super();
        this.wordBigram = wordBigram;
    }

    @Override
    protected int getEdgeWeight(int location, int vertex) {
        if(location >= 0) {
            IWord lastWord = getEdgeObject(location);
            IWord word = getEdgeObject(location, vertex);
            if(null != lastWord && null != word) {
                double conditionProb = getConditionProbability(lastWord.getWordName(), word.getWordName());
                if(conditionProb > 0.0001) {
                    conditionProb = -Math.log(conditionProb);
//                    System.out.println("Found Bigram: " + lastWord.getWordName() + " " + word.getWordName() + " prob: " + conditionProb);
                    return (int)((conditionProb / 3) + super.getEdgeWeight(location, vertex) * 2 / 3);
                }
            }
        }
        return super.getEdgeWeight(location, vertex);
    }

    protected double getConditionProbability(String word1, String word2) {
        return wordBigram.getProbability(word1, word2);
    }

    protected IWord getEdgeObject(int tail) {
        if(getDijkstraElement().hasFoundShortestPathTo(tail)) {
            int head = getNearestNeighbor(tail);
            return getEdgeObject(head, tail);
        }
        return null;
    }

    private IWord getEdgeObject(int location, int vertex) {
        return getGraph().getEdgeObject(location, vertex);
    }

    private int getNearestNeighbor(int end) {
        return getRouteBackTrace()[end];
    }
}

File: lib-segment/src/main/scala/websiteschema/mpsegment/graph/BigramDijkstra.scala
package websiteschema.mpsegment.graph

import websiteschema.mpsegment.dict.IWord

class BigramDijkstra(numOfVertexes: Int, wordBigram: WordBigram) extends DijkstraImpl(numOfVertexes: Int) {

    override def getEdgeWeight(location: Int, vertex: Int) : Int = {
        if(location >= 0) {
            val lastWord = getEdgeObject(location)
            val word = getEdgeObject(location, vertex)
            if(null != lastWord && null != word) {
                var conditionProb = getConditionProbability(lastWord.getWordName(), word.getWordName())
                if(conditionProb > 0.0001) {
                    conditionProb = -Math.log(conditionProb)
//                    println("Found Bigram: " + lastWord.getWordName() + " " + word.getWordName() + " prob: " + conditionProb)
                    return ((conditionProb / 3) + super.getEdgeWeight(location, vertex) * 2 / 3).toInt
                }
            }
        }
        return super.getEdgeWeight(location, vertex)
    }

    def getConditionProbability(word1: String, word2: String) : Double = {
        return wordBigram.getProbability(word1, word2)
    }

    def getEdgeObject(tail: Int) : IWord = {
        if(getDijkstraElement().hasFoundShortestPathTo(tail)) {
            val head = getNearestNeighbor(tail)
            return getEdgeObject(head, tail)
        }
        return null
    }

    private def getEdgeObject(location: Int, vertex: Int) : IWord = {
        return getGraph().getEdgeObject(location, vertex)
    }

    private def getNearestNeighbor(end: Int) : Int = {
        return getRouteBackTrace()(end)
    }
}
